### PR TITLE
mgr/dashboard: remove daemon list for badges

### PR DESCRIPTION
**Pull request by Pegonzal**
Signed-off-by: Pedro Gonzalez Gomez <pegonzal@redhat.com>
Fixes: https://tracker.ceph.com/issues/55895

Changes the UI for the rbd-mirroring daemons, from a table to badges color coded depending the daemon health property.

https://user-images.githubusercontent.com/106810838/172598700-385a4e63-181c-42ec-83ce-d56b67f43dda.mp4

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/overview/overview.component.html [hunk 1]
```diff
@@ -10,14 +10,14 @@
 </div>
 
 <div class="row">
-  <div class="col-sm-6">
-    <legend i18n>Daemons</legend>
+  <div class="col-md-12">
+    <strong i18n>Daemons: </strong>
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/overview/overview.component.html:14]:
```suggestion
    <strong i18n>rbd-mirror daemons: </strong>
```

**Review by pereman2 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.ts [hunk 2]
```diff
@@ -18,45 +17,55 @@ export class DaemonListComponent implements OnInit, OnDestroy {
...
   }
 
-  refresh() {
-    this.rbdMirroringService.refresh();
+  countDaemons(): {} {
+    let n = {unknown: 0, error: 0, warning: 0, success: 0};
+    this.empty = (this.data.length > 0) ? false : true; 
+    for (let i = 0; i < this.data.length; i++) {
+      let d = this.data[i];
+      if(d['health'] == 'OK')
```
[the remark this answers is not in the document]
Comment (pereman2) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.ts:48]: This function should have a unit test also.

**Comment by Sarthak0702** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.ts [hunk 3]
```diff
@@ -17,46 +16,55 @@ export class DaemonListComponent implements OnInit, OnDestroy {
...
+
+  daemonStatus(daemon_status: string): string {
+    if (daemon_status === 'success') {
+      return 'UP';
+    } else if (daemon_status === 'error') {
+      return 'DOWN';
+    } else if (daemon_status === 'warning') {
+      return 'WARNING';
+    } else {
+      return 'UNKNOWN';
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.ts:67]:
```suggestion
  daemonStatus(daemon_status: string): string {
    const display_names = {success: 'UP', error: 'DOWN', warning: 'WARNING', Unkown: 'UNKNOWN'};
    return display_names[daemon_status];
```

**Review by pereman2 — CHANGES_REQUESTED** [review]
nits

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.html [hunk 4]
```diff
@@ -1,13 +1,11 @@
...
-<ng-template #healthTmpl
-             let-row="row"
-             let-value="value">
-  <span [ngClass]="row.health_color | mirrorHealthColor">{{ value }}</span>
-</ng-template>
+<ng-container *ngFor="let daemon of daemons | keyvalue">
+  <span *ngIf="daemon.value > 0"
+        [ngClass]="daemon.key | mirrorHealthColor"
+        class="m-1"
+        i18n>{{daemon.value}} {{daemonStatus(daemon.key)}}</span>
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.html:10]:
```suggestion
        i18n>{{daemon.value}} {{daemon.key | daemonStatus}}</span>
```
We could use a [pipe](https://angular.io/guide/pipes) instead.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.ts [hunk 5]
```diff
@@ -17,46 +16,55 @@ export class DaemonListComponent implements OnInit, OnDestroy {
...
+      const daemon_data = this.data[i];
+      if (daemon_data['health'] === 'OK') {
+        daemon_states.success++;
+      } else if (daemon_data['health'] === 'Error') {
+        daemon_states.error++;
+      } else if (daemon_data['health'] === 'Warning') {
+        daemon_states.warning++;
+      } else {
+        daemon_states.unknown++;
+      }
```
[the remark this answers is not in the document]
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.ts]: Updated the function to use a dictionary and keep the same funcionality

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.html (hunk 4 above)
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.html]: Changed it to a pipe

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.html [hunk 6]
```diff
@@ -1,13 +1,11 @@
...
-<ng-template #healthTmpl
-             let-row="row"
-             let-value="value">
-  <span [ngClass]="row.health_color | mirrorHealthColor">{{ value }}</span>
-</ng-template>
+<ng-container *ngFor="let daemon of daemons | keyvalue">
+  <span *ngIf="daemon.value > 0"
+        [ngClass]="daemon.key | mirrorHealthColor"
+        class="m-1"
+        i18n>{{daemon.key | daemonStatus}}</span>
```
Comment (pereman2) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.html:10]: Did you test this?

that is how a pipe is used but `daemonStatus` itself isn't a pipe, it is a function. 
image
Check shared/pipes to see how they are implemented.
image

**Review by pereman2 — CHANGES_REQUESTED** [review]

**Comment by epuertat** [comment]
Interesting clean-up @Pegonzal !

image

This led me to think about a couple of things:
* In general Dashboard pages should be simple and following a single component/pattern: table, form, multi-card layout (landing page), etc. Why? Because each pattern serves a single purpose or concern: the table is for displaying a collection of items and their status/details, the form is for creating/editing each item settings, etc. However in the rbd-mirroring page I still see different purposes:
  * Connecting-configuring multiple clusters (bootstrapping)
  * Listing the rbd-mirroring daemons and status (but nothing can be changed there, so shouldn't this be on the services section?).
  * Listing the RBD pools (but RBD pools are also listed in the Pools section...)
  * Listing the mirror-enabled RBD images (but RBD images are also listed in the Block section...)
  * Rather than displaying multiple tables in a page, the tabs allow users to navigate through them, but only showing 1 table at a time.
* From a user perspective, what's their workflow? They probably had their RBD pool configured, with tons of RBD images serving as disks for virtual machines, and they now want to ensure that if this Ceph cluster becomes unavailable, the VMs will be able to fail over to other Ceph cluster. So:
  * They want to easily sync this current cluster to a newer one (the bootstrap process fulfills this)
  * They would want to receive alerts if something is not working or is underperforming: rbd images out of sync or not table to keep the pace (no rbd-mirror available, bandwidth or IOPS bottlenecks, split brain situations, etc.)
  * If that happens, they might want to have a deeper look at it and check the RBD image stats (progress, syncing status, etc).

So all the above considered:
* Now if no rbd-mirroring daemon and/or no RBD pools, a new empty state page will be displayed prompting to click on a button that configures the pre-requisites (RBD pool and rbd-mirror service).
* Then, we should immediately display the import-export form (this probably calls for a Wizard)
* After that, what should we display? I'd say: just a single table, probably the list of mirrored RBD images, pools and daemons:
image
* We have some RBD mirror metrics exported to Prometheus (not sure if all that we need, and more will come with the new exporter and per-image metrics), but none of them is used in the `monitoring/ceph-mixins/prometheus_alerts.yml` to trigger alerts, so perhaps we should add some:
```
ceph_prioritycache_cache_bytes{ceph_daemon="rbd-mirror.4203"} 2845415832.0
ceph_prioritycache_heap_bytes{ceph_daemon="rbd-mirror.4203"} 15409152.0
ceph_prioritycache_mapped_bytes{ceph_daemon="rbd-mirror.4203"} 15114240.0
ceph_prioritycache_target_bytes{ceph_daemon="rbd-mirror.4203"} 4294967296.0
ceph_prioritycache_unmapped_bytes{ceph_daemon="rbd-mirror.4203"} 294912.0
ceph_objecter_op_active{ceph_daemon="rbd-mirror.4203"} 0.0
ceph_objecter_op_r{ceph_daemon="rbd-mirror.4203"} 2.0
ceph_objecter_op_rmw{ceph_daemon="rbd-mirror.4203"} 0.0
ceph_objecter_op_w{ceph_daemon="rbd-mirror.4203"} 0.0
# HELP ceph_rbd_mirror_replay Replays
# TYPE ceph_rbd_mirror_replay counter
ceph_rbd_mirror_replay{ceph_daemon="rbd-mirror.4203"} 0.0
# HELP ceph_rbd_mirror_replay_bytes Replayed data
# TYPE ceph_rbd_mirror_replay_bytes counter
ceph_rbd_mirror_replay_bytes{ceph_daemon="rbd-mirror.4203"} 0.0
# HELP ceph_rbd_mirror_replay_latency_sum Replay latency Total
# TYPE ceph_rbd_mirror_replay_latency_sum counter
ceph_rbd_mirror_replay_latency_sum{ceph_daemon="rbd-mirror.4203"} 0.0
# HELP ceph_rbd_mirror_replay_latency_count Replay latency Count
# TYPE ceph_rbd_mirror_replay_latency_count counter
ceph_rbd_mirror_replay_latency_count{ceph_daemon="rbd-mirror.4203"} 0.0
# HELP ceph_rbd_mirror_snapshot_replay_bytes Replayed data
# TYPE ceph_rbd_mirror_snapshot_replay_bytes counter
ceph_rbd_mirror_snapshot_replay_bytes{ceph_daemon="rbd-mirror.4203"} 0.0
# HELP ceph_rbd_mirror_snapshot_snapshots Snapshots
# TYPE ceph_rbd_mirror_snapshot_snapshots counter
ceph_rbd_mirror_snapshot_snapshots{ceph_daemon="rbd-mirror.4203"} 0.0
# HELP ceph_rbd_mirror_snapshot_snapshots_time_sum Snapshots time Total
# TYPE ceph_rbd_mirror_snapshot_snapshots_time_sum counter
ceph_rbd_mirror_snapshot_snapshots_time_sum{ceph_daemon="rbd-mirror.4203"} 0.0
# HELP ceph_rbd_mirror_snapshot_snapshots_time_count Snapshots time Count
# TYPE ceph_rbd_mirror_snapshot_snapshots_time_count counter
ceph_rbd_mirror_snapshot_snapshots_time_count{ceph_daemon="rbd-mirror.4203"} 0.0
```
* All the alerts we are now triggering are pure Dashboard-generated notifications (in the absence of Prometheus alerts, better something than nothing). I like the red-orange badge thing, but we should ensure it gives user a hint on what's going on and how to fix it (the following badge doesn't have any description/tooltip):
image

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.html (hunk 6 above)
Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.html]: We [have a pipe](https://www.npmjs.com/package/ngx-pipe-function) that allows to call any (pure) function.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.ts [hunk 7]
```diff
@@ -17,46 +16,40 @@ export class DaemonListComponent implements OnInit, OnDestroy {
...
+    this.empty = this.data.length > 0 ? false : true;
+    for (let i = 0; i < this.data.length; i++) {
+      const health_color = this.data[i]['health_color'];
+      daemon_states[health_color]++;
+    }
+    return daemon_states;
+  }
+
+  daemonStatus(daemon_status: string): string {
+    const display_names = {success: 'UP', error: 'DOWN', warning: 'WARNING', info: 'UNKNOWN'};
```
[the remark this answers is not in the document]
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.ts:52]: Yo are right, I think it should be 'error'.

I also agree that info = unknoen is a little bit wierd but thats what I'm getting from the backend. From get_daemon_health function at src/pybind/mgr/dashboard/controllers/rbd_mirroring.py

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.ts [hunk 8]
```diff
@@ -17,46 +16,40 @@ export class DaemonListComponent implements OnInit, OnDestroy {
 
   subs: Subscription;
 
-  data: [];
-  columns: {};
+  data: any[];
+  daemons: {};
```
[the remark this answers is not in the document]
Comment (Pegonzal) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/mirroring/daemon-list/daemon-list.component.ts:20]: I Will do it. I'll be more careful with the naming

**Comment by epuertat** [comment]
> Thanks for your comments @epuertat! I have some more questions and coments regarding them:
> 
> Since this PR was ment to clean up some of the information regarding the daemons, and display the daemons health status in a more visual and simpler way should we tackle the next improvements in other issues/PRs?

Sure, feel free to take/drop the suggestions.

> > * Listing the rbd-mirroring daemons and status (but nothing can be changed there, so shouldn't this be on the services section?).
> 
> Couldn't that be a problem since we need an orchestrator to see that information from the services section?

While technically we could get some service information without orchestrator (Ceph tracks internally the status of Ceph services) you're right that this page won't work without orchestrator.


> 
> > * In general Dashboard pages should be simple and following a single component/pattern: table, form, multi-card layout (landing page), etc. Why? Because each pattern serves a single purpose or concern: the table is for displaying a collection of items and their status/details, the form is for creating/editing each item settings, etc. However in the rbd-mirroring page I still see different purposes:
> 
> That was the goal, to make it simple. A table is a great way to display information, but it seemed as there was not enough meaningful information inside the Daemons table to justify a whole table when there are two more tables displaying information aswell. There might be too much going at the same time, which most of the time might not be needed.

My idea of simplification here is more about adhering to generic patterns and removing ad-hoc (per-component) customizations, so the goal would be that every Dashboard page is just a table + form component. Anything else that adds specificities to a page would 'complicate' it. If we find that these new daemon labels make sense for ALL components, yes, let's go for it. But again, it's a suggestion. 

> 
> > * Rather than displaying multiple tables in a page, the tabs allow users to navigate through them, but only showing 1 table at a time.
> 
> This could be a great solution overall. However I think two main issues might rise from it. Right now the Images table already has three different tabs, so how could we manage that with the proposed change to tabs. And again, would it be really necesary a whole table to display the daemons data. Is it something the user could be interested?

Agree on the daemons table. Tables should be displaying a medium-large amount of data. For just displaying 1-2 daemons it makes it a waste of space.

> > * Now if no rbd-mirroring daemon and/or no RBD pools, a new empty state page will be displayed prompting to click on a button that configures the pre-requisites (RBD pool and rbd-mirror service).
> 
> Would this PR cover that #46527 ?
> 

If there are no daemons, then the "welcome screen" would replace this page.

**Comment by Pegonzal** [comment]
Then, after our discussion and your comments @epuertat my conclusion would be as follows: 

This PR is a great implementation because it can show relevant information (the daemon status) to the user in a clear and simple way at everymoment at the top of the page but this doesn't follow the simplicity of following a single component/pattern we are looking for with our Dashboard pages. **So it should either be discarded or added to other pages where might be needed to keep the consistency but adding another pattern witch will reduce the simplicity of the pages.** So which of the two options would be the best?

However is it clear for all of us that this page would be beneffited from some kind of cleanup. Then the next things could be added/modified to do so:

- **if no rbd-mirroring daemon and/or no RBD pools, a new empty state page will be displayed prompting to click on a button that configures the pre-requisites (RBD pool and rbd-mirror service).**
- **Then, we should immediately display the import-export form (through a Wizard)**
- **Display a single table, the list of mirrored RBD images, pools and daemons:**
image
- **Add Prometheus metrics to trigger alerts**
- **Add hints for the user to know whats going on** 

@pereman2 since you were also in favor of this PR, I would also like to know your thoughs about this conclusions.

Thank you guys for your help and comments and I'll wait for your responde about this.

**Comment by pereman2** [comment]
@epuertat I think this is a problem of effectively displaying relevant data. I understand that you want to have consistency across components and try not to do ad-hoc implementations.
In this case we don't have single data source in rbd-mirroring like in pools, osds, hosts... The closest thing would be RGW where wwe have a link to a daemons table where we see a lot of information related to that **rgw daemon**, but that's the thing, there is relevant info in that table. **Relevant info includes a whole list of details, realms, performance counters**...

On the other side, we can agree that **rbd-mirror** is a different beast which works in a peculiar way. In the daemons table the **relevant info is: name maybe and the state of the daemon** because you cannot opererate if all dameons are down or with issues or whatever. If you were about to do something on an mirrored image and all daemons are down, you'd prefer not having to go to another tab before performing another action or checking if something is happening with daemons. If we have the same layout as we've always had, it wouldn't be a problem but still, you would have a huge table taking unnecessary space. With the current approach in this PR I believe you can have the same info the table provides, in a compact way without having to context switch between tables which is something I think it's important.

If your problem is consistency and that we would have to implement this everywhere next time, I don't think we should not change something because it doesn't stick to our usual bloated unreadable tables,  we should strive to display the data in the most efficient way possible and it has to be relevant.

**Review by nizamial09 — COMMENTED** [review]
While I agree that the rbd-mirroring page needs to be improved a lot, we should still be following some UX patterns. I know that we are kind of limited in that area but atleast we can try to follow the general theme and improve upon it. Our goal here in the end is to show what is relevant to the user upfront. And while showing the `Daemons` table in the first view is not mostly relevant, we can still show it. Just not in the front.

I think the Images (or the status of the images) is the important thing here and one that user will continue to come to look at even after setting rbd-mirror up. So just show the info in the Images tab first and then do everything in a different tab would make a lot of sense here. It'll adhere to the UX we have in place too. 

While the Daemons page doesn't give a lot of info, it still shows which daemon is up and which is not. I don't like the idea of this being lost under a tooltip TBH.

Picture speaks a thousand words here so my vision is to have something like this in place?
Screenshot from 2022-06-29 20-40-22

I really don't want to lose this opportunity to do some cleanups around this page. So lets try to bring a good design that we all can agree on.

Let me know your thoughts on this too. @Pegonzal @epuertat @pereman2

**Review by pereman2 — COMMENTED** [review]
@nizamial09 I'm ok with that approach. The only concern with is having again 3 tables displayed in a row manner. Feels weird.

**Comment by epuertat** [comment]
I agree with @pereman2. We should find a way to display 1 piece of information at once (that's how some view abstractions work: pages, tabs) and avoid showing too much information simultaneously (collapsible/accordion components are acceptable, but risky if abused).

I'm fine with any improvement that reduces/condensates the amount of information displayed.

My comment today about adding badges/labels/chips (whichever we call them) is that we should prefer highlighting information that requires attention:
* Using a red label for the "critical" alert label in the alerts table column is better than not highlighting anything. 
* Using a red label for displaying the count of critical alerts in the tab header, menu entry and similar places is even better (it aggregates information and directs the user to the place where the detailed info is).
* Using the labels to display the number of UP/DOWN servers is ok too (better than the non highlighted version), but I had 2 issues with that: the placement was not clear (floating between the breadcrumbs and the tabs), and if we can detect the undesired situation (1 server DOWN), why not reporting that instead of forcing the user to digest and process both the ok info (X servers UP) and not-ok info (X servers DOWN)?

**Pull request closed without merging**